components/renku_data_services/users/db.py

```diff
@@ -124,10 +124,35 @@ async def _get_users(self, email: str | None = None) -> list[UserInfo]:
 
             return [user.dump() for user in users if user.namespace is not None]
 
+    async def remove_user(self, user_id: str) -> str | None:
+        """Remove a user."""
+        logger.info(f"remove_user: Trying to remove user with ID {user_id}")
+        return await self._remove_user(user_id=user_id)
+
+    @with_db_transaction
+    @dispatch_message(avro_schema_v2.UserRemoved)
+    async def _remove_user(self, user_id: str, *, session: AsyncSession | None = None) -> str | None:
+        """Remove a user from the database."""
+        if not session:
+            raise errors.ProgrammingError(message="A database session is required")
+        logger.info(f"Trying to remove user with ID {user_id}")
+        stmt = delete(UserORM).where(UserORM.keycloak_id == user_id).returning(UserORM)
+        user = await session.scalar(stmt)
+        await self.authz._remove_user_namespace(user_id)
```

Comment:
The original function may have been written before `@Authz.authz_change` was available.

It should be placed just below `@with_db_transaction`, for example:
```python
@with_db_transaction
@Authz.authz_change(AuthzOperation.delete, ResourceType.user)
@dispatch_message(avro_schema_v2.UserRemoved)
async def _remove_user(self, user_id: str, *, session: AsyncSession | None = None) -> str | None:
    [...]
```

Then the corresponding changes will need to exist in `components/renku_data_services/authz/authz.py`:
* Add cases in `_get_authz_change()`

Potentially open for discussion is the fact that the current `authz` changes only remove the user namespace but do not remove any permission associated with that user. I think this should probably be done, otherwise the user will still appear in membership lists.

Reply:
I confirmed with @olevski that the current code does not remove all references of that user in the `authz` database, but we should do that (since the user ID would still be listed in groups and projects).

@wesjdj I can help update the code in `authz.py` if needed.

Reply:
@leafty Thanks! I've done the initial change requested now. We'd need to look up the references in the spicedb first, I assume?

Reply:
Makes sense! Should we wait on merging this PR then until #425 has been merged?

Reply:
I can also write a `_remove_user()` for this PR.

Reply:
@wesjdj Code is available in https://github.com/SwissDataScienceCenter/renku-data-services/pull/427.